### PR TITLE
Add fast_beam_search support

### PR DESCRIPTION
**Pull request by pkufool**
Only support decoding with trivial_graph, and I only added to streaming conformer server, will add it to other models.

~~Needs https://github.com/k2-fsa/k2/pull/1018~~

### CODE REVIEW

**Comment by pkufool** [comment]
@csukuangfj  What's your plan for handling fsa attributes in sherpa, I think it is better to do attributes propagation in C++ side, we can either pass this attributes (mainly aux_labels) to fuctions or move FsaClass in v2.0-pre to sherpa.

**Comment by danpovey** [comment]
I'm wary about adding too much complexity on the C++ level (especially that duplicates things we have implemented in Python) unless there are concrete reasons why we need it, e.g. it would have a noticeable impact on efficiency.  The C++ code is in general more of a burden to maintain because the pool of possible developers will be lower.

**Comment by pkufool** [comment]
> 

Ok, I see. let's keep only the core part of fsa algorithm on C++ level.  It is possible to reuse the k2 python code, I think attributes propagation won't take too much time.  What I was think of implementing it on C++ level is just make the C++ interface more clean and elegant (I.e. no need to pass out arc_map).

**Comment by danpovey** [comment]
Hm.  Bear in mind that we use arc_map for a lot of things, like backprop and propagating various kinds of attributes, so it's likely needed at some level.

**Comment by ezerhouni** [comment]
@pkufool Please ping me if I can be of any help !

**Comment by pkufool** [comment]
> @pkufool Please ping me if I can be of any help !

Thanks! I think most of the stuffs in Sherpa have finished now, we are deciding what is the best way to expose k2 interface.  One thing you can try is to implement fast_beam_search for other two models, current implementation might change though if we change the interface in k2.

**Comment by ezerhouni** [comment]
@pkufool I will try to find some time in the coming days. Would you rather me adding commits in your branch or creating a new PR ?

**Comment by pkufool** [comment]
> @pkufool I will try to find some time in the coming days. Would you rather me adding commits in your branch or creating a new PR ?

Thanks! For this PR, I think it is better to add commits to my branch, it would be too much hassle to fix conficts by creating a new PR.

File: sherpa/csrc/rnnt_beam_search.cc [hunk 1]
```diff
@@ -423,4 +417,85 @@ std::vector<std::vector<int32_t>> ModifiedBeamSearch(
...
+                              log_semiring, &entering_arcs);
+
+  k2::Ragged<int32_t> best_path_arc_indexes =
+      k2::ShortestPath(lattice, entering_arcs);
+
+  k2::FsaVec out = k2::FsaVecFromArcIndexes(lattice, best_path_arc_indexes);
+  return out;
+}
+
+std::vector<std::vector<int32_t>> FastBeamSearch(
```
Comment (ezerhouni) [inline sherpa/csrc/rnnt_beam_search.cc:439]: If I am not mistaken, this function corresponds to : ```fast_beam_search_one_best``` : https://github.com/k2-fsa/icefall/blob/master/egs/librispeech/ASR/pruned_transducer_stateless2/beam_search.py#L30
For clarity, it might be better to keep the same function as in icefall, to avoid any misunderstanding, what do you think ?

Comment (pkufool) [reply sherpa/csrc/rnnt_beam_search.cc:439]: Sorry, we changed our mind, and decided not to write any C++ code for fast_beam_search in sherpa. The idea is we will release the gil in k2's pybind code. Will update this PR soon.  Actually, this won't affect too much of your code already written. Thanks!

Comment (pkufool) [reply sherpa/csrc/rnnt_beam_search.cc:439]: Yes, we will implement sherpa `fast_beam_search_one_best` in pure python, almost the same as it in icefall.

**Comment by pkufool** [comment]
>  Could you rebase your branch with the current master so I can add fast beam search to conv_enformer ? 

Of course, will update the code as well as my recent change.

**Comment by pkufool** [comment]
Now, it doesn't depend https://github.com/k2-fsa/k2/pull/1018 anymore. 

@ezerhouni  I suggest to finish `fast_beam_search_one_best` for all the models, then you can implement `fast_beam_search_LG` in a new PR if you have time.

**Comment by pkufool** [comment]
@csukuangfj  Can you have a look at this PR and let's merge it as soon as possible. Then @ezerhouni  can continue his work in a new PR, I think it would be more convenient for him.

File: sherpa/bin/streaming_conformer_rnnt/decode.py [hunk 2]
```diff
@@ -68,13 +76,23 @@ def __init__(
...
+        if decoding_method == "fast_beam_search":
+            assert decoding_graph is not None
+            self.rnnt_decoding_stream = k2.RnntDecodingStream(decoding_graph)
+            self.hyp = []
+        elif decoding_method == "greedy_search":
+            assert decoder_out is not None
+            self.decoder_out = decoder_out
+            self.hyp = [blank_id] * context_size
+        else:
+            assert (
```
Comment (csukuangfj) [inline sherpa/bin/streaming_conformer_rnnt/decode.py:89]: Please use
```python3
raise ValueError(f"Decoding method {decoding_method} is not supported.")

File: sherpa/bin/streaming_conformer_rnnt/decode.py [hunk 3]
```diff
@@ -68,13 +76,23 @@ def __init__(
...
 
         self.states = initial_states
-        self.decoder_out = decoder_out
+        self.decoding_graph = decoding_graph
+        if decoding_method == "fast_beam_search":
+            assert decoding_graph is not None
+            self.rnnt_decoding_stream = k2.RnntDecodingStream(decoding_graph)
+            self.hyp = []
+        elif decoding_method == "greedy_search":
+            assert decoder_out is not None
```
Comment (csukuangfj) [inline sherpa/bin/streaming_conformer_rnnt/decode.py:85]: Please document that `decoder_out` is used only for `greedy_search`.

File: sherpa/bin/streaming_conformer_rnnt/decode.py [hunk 4]
```diff
@@ -57,6 +60,11 @@ def __init__(
           initial_states:
             The initial states of the Conformer model. Note that the state
             does not contain the batch dimension.
+          decoding_method:
+            The decoding method to use, currently, only greedy_search and
+            fast_beam_search are supported.
+          decoding_graph:
```
Comment (csukuangfj) [inline sherpa/bin/streaming_conformer_rnnt/decode.py:66]: Please document that it cannot be `None` for fast_beam_search decoding.

File: sherpa/bin/streaming_conformer_rnnt/streaming_server.py [hunk 5]
```diff
@@ -180,18 +219,26 @@ def run_model_and_do_greedy_search(
...
+        decoder_out_list = []
+        hyp_list = []
+    else:
+        rnnt_decoding_streams_list = []
+        rnnt_decoding_config = server.rnnt_decoding_config
     for s in stream_list:
+        if decoding_method == "greedy_search":
+            decoder_out_list.append(s.decoder_out)
+            hyp_list.append(s.hyp)
+        if decoding_method == "fast_beam_search":
```
Comment (csukuangfj) [inline sherpa/bin/streaming_conformer_rnnt/streaming_server.py:239]: The name of this function is `run_model_and_do_greedy_search`.

Please move it to a separate function, e.g.,
`run_model_and_do_fast_beam_search`

File: sherpa/bin/streaming_conformer_rnnt/streaming_server.py [hunk 6]
```diff
@@ -319,20 +395,37 @@ def __init__(
...
+            self.rnnt_decoding_config = k2.RnntDecodingConfig(
+                vocab_size=self.vocab_size,
+                decoder_history_len=self.context_size,
+                beam=beam,
+                max_states=max_states,
+                max_contexts=max_contexts,
+            )
+            self.decoding_graph = k2.trivial_graph(self.vocab_size - 1, device)
+        else:
```
Comment (csukuangfj) [inline sherpa/bin/streaming_conformer_rnnt/streaming_server.py:419]:
```suggestion
        elif decoding_method == "greedy_search":
           ...
         else:
             raise an exception
```

File: sherpa/bin/streaming_conformer_rnnt/streaming_server.py [hunk 7]
```diff
@@ -487,7 +582,7 @@ async def handle_connection_impl(
             while len(stream.features) > self.chunk_length:
                 await self.compute_and_decode(stream)
                 await socket.send(
-                    f"{self.sp.decode(stream.hyp[self.context_size:])}"
+                    f"{self.sp.decode(stream.hyp[self.context_size:] if self.decoding_method != 'fast_beam_search' else stream.hyp)}"
```
Comment (csukuangfj) [inline sherpa/bin/streaming_conformer_rnnt/streaming_server.py:585]: Please refactor it so that it is easier to add other decoding methods.

File: sherpa/csrc/rnnt_emformer_model.h [hunk 8]
```diff
@@ -72,6 +72,7 @@ class RnntEmformerModel : public RnntModel {
   int32_t BlankId() const override { return blank_id_; }
   int32_t UnkId() const override { return unk_id_; }
   int32_t ContextSize() const override { return context_size_; }
+  int32_t VocabSize() const override { return vocab_size_; }
```
Comment (csukuangfj) [inline sherpa/csrc/rnnt_emformer_model.h:75]: @yaozengwei Please configure your vim to format code automatically.

File: sherpa/python/sherpa/decode.py [hunk 9]
```diff
@@ -0,0 +1,209 @@
...
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import k2
+import torch
+from _sherpa import RnntModel
+from typing import List, Union
```
Comment (csukuangfj) [inline sherpa/python/sherpa/decode.py:19]: Please sort it using `isort`.

File: sherpa/bin/streaming_conformer_rnnt/decode.py [hunk 10]
```diff
@@ -15,8 +15,9 @@
 # limitations under the License.
 
 import math
-from typing import List
+from typing import List, Optional
```
Comment (csukuangfj) [inline sherpa/bin/streaming_conformer_rnnt/decode.py:18]: Please also update CI to test the newly added `fast_beam_search`.

You can use

https://github.com/k2-fsa/sherpa/pull/66/files#diff-0fc8407285d79be4141309a772e34bdb812ee4b6444e80f69191cbe0ad6d51a9

as an example to update the CI.

**Review by csukuangfj — CHANGES_REQUESTED** [review]

File: sherpa/bin/streaming_conformer_rnnt/streaming_server.py (hunk 5 above)
Comment (pkufool) [reply sherpa/bin/streaming_conformer_rnnt/streaming_server.py:239]: Oh, I did not notice this, will refactor it.

Comment (pkufool) [reply sherpa/bin/streaming_conformer_rnnt/streaming_server.py:239]: I just changed the function name from `run_model_and_do_greedy_search` to `run_model_and_do_search`, I think it would be too much duplicate code if creating a separate function.

File: sherpa/python/sherpa/decode.py [hunk 11]
```diff
@@ -0,0 +1,210 @@
...
+        )
+        log_probs = (logits / temperature).log_softmax(dim=-1)
+        decoding_streams.advance(log_probs)
+    decoding_streams.terminate_and_flush_to_streams()
+    lattice = decoding_streams.format_output(processed_lens.tolist())
+
+    return lattice
+
+
+def get_texts(
```
[the remark this answers is not in the document]
Comment (csukuangfj) [reply sherpa/python/sherpa/decode.py:153]: How about making it a dependency now? We can make `icefall` installable by `pip install icefall`.

[the remark this answers is not in the document]
Comment (ezerhouni) [reply sherpa/python/sherpa/decode.py:153]: Agree !

[the remark this answers is not in the document]
Comment (pkufool) [reply sherpa/python/sherpa/decode.py:153]: @csukuangfj  How about merging this PR first, then @ezerhouni  can continue this thing.

[the remark this answers is not in the document]
Comment (ezerhouni) [reply sherpa/python/sherpa/decode.py:153]: I can take care of it

[the remark this answers is not in the document]
Comment (pkufool) [reply sherpa/python/sherpa/decode.py:153]: @ezerhouni I think it is OK to keep a copy of  `get_texts` and `one_best_decoding` right now, only several lines of code. Let's make icefall as a dependency when we need more functions from icefall。

**Comment by pkufool** [comment]
@ezerhouni Merged, you can take over this task, thanks!
